core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlRelationBuilder.scala

```diff
@@ -64,17 +65,74 @@ class TSqlRelationBuilder(override val vc: TSqlVisitorCoordinator)
     case Some(errorResult) => errorResult
     case None =>
       ctx match {
-        case qs if qs.querySpecification() != null => qs.querySpecification().accept(this) // TODO: Implement set ops
+        case qs if qs.querySpecification() != null =>
+          val lhs = qs.querySpecification().accept(this)
+          buildIntersectOperations(lhs, qs.sqlUnion().asScala)
         case qe if qe.queryExpression() != null =>
-          qe.queryExpression().asScala.map(_.accept(this)) match {
-            case Seq(lhs) => lhs
-            case Seq(lhs, rhs) =>
-              val isAll = qe.ALL() != null
-              ir.SetOperation(lhs, rhs, ir.UnionSetOp, is_all = isAll, by_name = false, allow_missing_columns = false)
+          qe.queryExpression().asScala.map(_.accept(this)).reduceLeft { (lhs, rhs) =>
+            val isAll = qe.ALL() != null
+            ir.SetOperation(lhs, rhs, ir.UnionSetOp, is_all = isAll, by_name = false, allow_missing_columns = false)
           }
       }
   }
 
+  @tailrec
+  private[this] def buildIntersectOperations(
```

Comment:
I wonder if we could simplify that a bit by first going through the the seq of context and split it at each instance of `INTERSECT`, producing a `Seq[Seq[SqlUnionContext]]` that we'd next traverse again to produce the actual `SetOperation(..., IntersectSetOp)`.

Something along the lines of:

```scala
def appendNext(agg: Seq[Seq[SqlUnionContext]], item: SqlUnionContext) = Seq(item) +: agg
def appendFirst (agg: Seq[Seq[SqlUnionContext]], item: SqlUnionContext) = 
  if (agg.isEmpty) Seq(Seq(item))
  else (item +: agg.head) +:  agg.tail

remainingContexts.foldLeft(Seq.empty[Seq[SqlUnionContext]]) {
  case (agg, item) if item.INTERSECT() == null => appendFirst(item)
  case (agg, item) => appendNext(item)
}
.map( // here you get seqs of SqlUnionContext that aren't INTERSECT, in reverse order, so you may foldRight/reduceRight )
.reduceRight(//here you build your SetOperation(..., IntersectSetOp, ...))
```

Reply:
Having slept over it, maybe the existing @tailrec approach would still be more ... approachable. Your call.